From 83511b9ff0d9af4569a4579ad392894396fe6a79 Mon Sep 17 00:00:00 2001
From: Rob Clark <robdclark@chromium.org>
Date: Thu, 18 Jul 2024 14:16:48 -0700
Subject: [PATCH 13/17] freedreno/drm: Handle a7xx case

Signed-off-by: Rob Clark <robdclark@chromium.org>
---
 src/freedreno/drm/freedreno_pipe.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/src/freedreno/drm/freedreno_pipe.c b/src/freedreno/drm/freedreno_pipe.c
index 9b8abb7b602..ea14976545e 100644
--- a/src/freedreno/drm/freedreno_pipe.c
+++ b/src/freedreno/drm/freedreno_pipe.c
@@ -200,8 +200,17 @@ uint32_t
 fd_pipe_emit_fence(struct fd_pipe *pipe, struct fd_ringbuffer *ring)
 {
    uint32_t fence = ++pipe->last_fence;
-
-   if (pipe->is_64bit) {
+   unsigned gen = fd_dev_gen(&pipe->dev_id);
+
+   if (gen >= A7XX) {
+      OUT_PKT7(ring, CP_EVENT_WRITE7, 4);
+      OUT_RING(ring, CP_EVENT_WRITE_0_EVENT(CACHE_FLUSH_TS) |
+               CP_EVENT_WRITE7_0_WRITE_SRC(EV_WRITE_USER_32B) |
+               CP_EVENT_WRITE7_0_WRITE_DST(EV_DST_RAM) |
+               CP_EVENT_WRITE7_0_WRITE_ENABLED);
+      OUT_RELOC(ring, control_ptr(pipe, fence));   /* ADDR_LO/HI */
+      OUT_RING(ring, fence);
+   } else if (gen >= A5XX) {
       OUT_PKT7(ring, CP_EVENT_WRITE, 4);
       OUT_RING(ring, CP_EVENT_WRITE_0_EVENT(CACHE_FLUSH_TS));
       OUT_RELOC(ring, control_ptr(pipe, fence));   /* ADDR_LO/HI */
-- 
2.45.2

